tests/customlists/test_export.py

```diff
@@ -0,0 +1,290 @@
+import json
+import logging
+from pathlib import Path
+
+import pytest
+
+from customlists.customlist_export import (
+    CustomListExporter,
+    CustomListExportFailed,
+    CustomListExports,
+)
+from tests.core.util.test_mock_web_server import MockAPIServer, MockAPIServerResponse
+
+
+@pytest.fixture
+def mock_web_server():
+    """A test fixture that yields a usable mock web server for the lifetime of the test."""
+    _server = MockAPIServer("127.0.0.1", 10256)
+    _server.start()
+    logging.info(f"starting mock web server on {_server.address()}:{_server.port()}")
+    yield _server
+    logging.info(
+        f"shutting down mock web server on {_server.address()}:{_server.port()}"
+    )
+    _server.stop()
+
+
+class TestExports:
+    @staticmethod
+    def _customlists_resource_path(name) -> str:
+        """The path to the customlists resource with the given filename."""
+        base_path = Path(__file__).parent.parent.parent
+        resource_path = base_path / "customlists"
+        return str(resource_path / name)
+
+    @staticmethod
+    def _test_customlists_resource_path(name) -> str:
+        """The path to the customlists test resource with the given filename."""
+        base_path = Path(__file__).parent.parent
+        resource_path = base_path / "customlists" / "files"
+        return str(resource_path / name)
+
+    @staticmethod
+    def _test_customlists_resource_bytes(name) -> bytes:
+        with open(TestExports._test_customlists_resource_path(name), "rb") as f:
+            return f.read()
+
+    def test_export_auth_fails(self, mock_web_server: MockAPIServer, tmpdir):
+        """If the server returns a ~400 error code, signing in fails."""
+        sign_response = MockAPIServerResponse()
+        sign_response.status_code = 401
+        mock_web_server.enqueue_response(
```

Comment:
Rather than start up a server for the test cases can we `mock.patch` the appropriate functions? It tends to be more efficient, consequently allowing for quicker test execution

Reply:
I mean, we can... It's a question of taste, really. I don't like patching because it touches internals that are subject to change, and it often makes the tests difficult to understand. There's a slight conceptual difference between testing the side effects that the code produces (the requests it makes to the server), and testing the internal calls it makes (by checking that it makes the calls to the requests library that we think it should).

Reply:
Well here we're testing everything all the way down to the requests library. The philosophy of patching is that we patch whatever we don't care about in the test. In this case we can stop once we exit the class and start using the internal http.py classes. We just need to know we are using those correctly and working with the expected responses, so if we patch the http.get_with_timeout function (or something before that) we know we're using it correctly and using its expected responses correctly, without actually making network calls.

Reply:
I've been giving this a lot of thought, and I think that's the problem: By patching, someone reading the tests later has to then go and search through the internals of the code to see what specific set of parameters are being passed to the `requests` library in order to fix a test that might now be failing due to changes. It trades clarity of code for execution efficiency, and it requires anyone reading the tests to know very specific low level internal details. This has a real cost in terms of the time it takes to work on code.

A good example was the recent Overdrive work: I ended up spending all day trying to track down why a test was failing, because the actual HTTP requests were buried multiple layers deep in the code. If instead the tests had been written in a way where I only needed to observe the external outgoing requests (the _side effects_ as opposed to the specifics of how calls were made internally), I'd have been done in half an hour.

Reply:
This is a fair counter point